webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py

```diff
@@ -0,0 +1,224 @@
+            'context': new_tab["context"],
+            'navigation': result["navigation"],
+            'timestamp': any_int,
+            'url': url(EMPTY_PAGE)
+        },
+        await on_frame_navigated,
+    )
+
+
+@pytest.mark.parametrize(
+    "navigation_kind",
+    ["browsing_context.navigate", "script.call_function"],
+)
+@pytest.mark.parametrize(
+    "hash_before, hash_after",
+    [
+        ("", "#foo"),
+        ("#foo", "#bar"),
+        ("#foo", "#foo"),
+        ("#bar", ""),
+    ],
+    ids=[
+        "without hash to with hash",
+        "with different hashes",
+        "with identical hashes",
+        "with hash to without hash",
+    ],
+)
+async def test_history_api(
+    bidi_session, new_tab, url, subscribe_events, wait_for_event, hash_before, hash_after, navigation_kind
+):
+    await bidi_session.browsing_context.navigate(
+        context=new_tab["context"], url=url(EMPTY_PAGE + hash_before), wait="complete"
+    )
+
+    await subscribe_events([FRAGMENT_NAVIGATED_EVENT])
+
+    on_frame_navigated = wait_for_event(FRAGMENT_NAVIGATED_EVENT)
+
+    if navigation_kind == "browsing_context.navigate":
+      result = await bidi_session.browsing_context.navigate(
+          context=new_tab["context"], url=url(EMPTY_PAGE + hash_after))
+
+      recursive_compare(
+          {
+              'context': new_tab["context"],
+              'navigation': result["navigation"],
+              'timestamp': any_int,
+              'url': url(EMPTY_PAGE + hash_after)
+          },
+          await on_frame_navigated,
+      )
+    elif navigation_kind == "script.call_function":
+      await bidi_session.script.call_function(
+          raw_result=True,
+          function_declaration="""(url) => {
+              history.pushState(null, null, url);
+          }""",
+          arguments=[
+              {"type": "string", "value": url(EMPTY_PAGE + hash_after)},
+          ],
+          await_promise=False,
+          target=ContextTarget(new_tab["context"]),
+      )
+
+      recursive_compare(
+          {
+              'context': new_tab["context"],
+              'navigation': None,
+              'timestamp': any_int,
+              'url': url(EMPTY_PAGE + hash_after)
+          },
+          await on_frame_navigated,
+      )
+    else:
+       raise Exception("Unknown navigation_kind")
+
+
+@pytest.mark.parametrize("type_hint", ["tab", "window"])
+async def test_new_context(bidi_session, subscribe_events, wait_for_event, type_hint):
+    await subscribe_events(events=[FRAGMENT_NAVIGATED_EVENT])
+
+    events = []
+
+    async def on_event(method, data):
+        events.append(data)
+
+    remove_listener = bidi_session.add_event_listener(FRAGMENT_NAVIGATED_EVENT, on_event)
+
+    new_context = await bidi_session.browsing_context.create(type_hint=type_hint)
+
+    wait = AsyncPoll(bidi_session, timeout=0.5)
+    with pytest.raises(TimeoutException):
+        await wait.until(lambda _: len(events) > 0)
+
+    remove_listener()
```

Comment:
I assume that this event also needs to be fired for `document.write` usage? Given that we had a problem in the past we should potentially add that as well. This would also include the test for a base tag which both other events already have.

Reply:
I think that there should be no navigation event for document.write usage. I will add a test to check that. As for the base tag, I am not sure how it related to the navigation event? Was it added because of an issue in the Firefox implementation?

Reply:
As we discussed some months ago and also implemented in Firefox the `document.write` usage should trigger the `DOMContentLoaded` and `load` events. So not sure if I understand your reply. Did you only refer to the `browsingContext.fragmentNavigated` event?

Yes, for URLs with a base tag we incorrectly reported the URL. Don't you think that such a test is helpful for other browsers as well?

Reply:
Yes, I mean that `document.write` should not trigger `browsingContext.fragmentNavigated` and I have added a test. I have also added the base tag test yesterday (thanks for clarifying that the test exists due to a bug).

Reply:
Ok, so I'm still a bit baffled and sorry for mentioning that so late. In the above tests (like `test_timestamp`) which exercise a normal navigation from eg. domain A to domain B the tests also expect a `fragmentNavigated` event. Why that? As per spec the event should only be fired when a navigation within the same page is happening but those navigations don't qualify for that.

Reply:
Ah, good catch! Apparently it is a bug in the implementation! I will update the test.

Reply:
Updated the test and added a test that regular navigations should not emit this event.